add deactivate and reregister operator interfaces (#70)
subspace/protocol-specs: docs/decex/interfaces.md

Initiates an epoch transition immediately, without waiting for a `StakeEpochDuration` blocks to be created. 
This operation can only be initiated by a root user. It is used to ensure domain liveness under circumstances when, for example, a large amount of stake went offline, and a domain can’t produce blocks frequently enough.

### deactivate_operator  
  

> `deactivate_operator(operator_id)`  

Reviewer: I think a constant cooldown period is fine for now, but we might want to allow an optional epoch count parameter in future. For example, if we know an operator is going to be offline for a shorter or longer time.

Reviewer 2: The initial intention was to be more reactive (i.e. an operator is not producing bundles and is impacting chain liveness, so we should deactivate them) vs proactively working with an operator that knows they will be down for some time. Maybe that is a use case I hadn't thought of, it could also be handled by allowing an operator to deactivate themselves. Regarding a cooldown period, it feels like added complexity and tracking with limited benefit.

Author: The reason cooldown is added to be bit more flexible when protocol itself deactivates the user based on their offliness and having a cooldown would be ideal to ensure operator immediately tries to rejoin.

At the current state, it is explicit sudo call but down the line reactivation is can be done by the operator themselves

Reviewer: Ok, in that case I think it's a good idea to experiment with a fixed cooldown, to see if it has a useful value, or if it needs to be dynamic.

Reviewer 2: One nit that @vedhavyas and I discussed today was the naming of cooldown period (naming is hard!). I don't think it really describes what it is well. I think something like `reactivation_delay`or`min_reactivation_wait_period` would be more descriptive.

>   

Reviewer: Sudo or governance can deactivate all the operators. Should this be possible? Or should we add a check that there are 2 or more operators before starting to deactivate one of them?

Similar question for deregistering operators, but that's outside the scope of this PR.

if we decide to allow it, let's document that somewhere?

Author: fair. I think we should ensure there is atleast one operator in the current and future set where we would not allow such an operation.

But it is tricky to handle since there is an edge case
- two operators are running
- deactivate one operator
- the remaining operator decides to deregister themselves
- Then technically we should not allow them since Domains will not progress from the next epoch

This needs a bit more thought in when deactivating operator

Reviewer: That's correct, the rule would be "one or more active operators".

I don't think it needs to be fixed in this PR, but I'd like it to be documented that it's possible to deactivate or deregister all the operators.

There might also be cases where we want to do that, if we want to effectively end or pause a domain.

Reviewer 2: I think for the first phase this would require too much complexity.

> Sudo or Governance can deactivate a given operator from the Registered status. This can be done if the operator is  

Reviewer 2: Should an operator also be able to `deactivate` themselves. This could be useful if they know they will need to bring down their node for some period of time or if they are experiencing technical issues and they do not want to impact the liveness of the chain but they also don't want to deregister and lose all nominated stake.

Author: Fair point. Can be included

Reviewer 2: @vedhavyas after our discussion today let's hold off on this as it may add too much complexity for phase 1. Specifically, the complexity is added around who can trigger a reactivation after an operator driven deactivation.

Reviewer 3: I would argue that we don't necessarily need the ability to self deactivate. Operators should be aiming to have redundancy & resilience in place such that they are able to maintain their operation with zero or minimal downtime.

> offline for any given reason and the block production on Domains significantly reduced due to the stake deposited for the given operator. By deactivating the operator, the operator status is moved from `Registered` to `Deactiavted` and operator is removed from the Domain operator set and next Operator set along with their Stake in the Domain. Rest of the operators will continue producing bundles and Domain block production can be restored since the offline operator's stake is not pledged to Domain anymore.

Reviewer 2: A couple of grammatical nits.
```suggestion
offline for any given reason and the block production on Domains is significantly reduced due to the stake deposited for the given operator. By deactivating the operator, the operator status is moved from `Registered` to `Deactiavted` and operator is removed from the Domain operator set and next Operator set along with their stake in the Domain. The rest of the operators will continue producing bundles and expected Domain block production can be restored since the offline operator's stake is not pledged to Domain anymore.
```

> 

Reviewer: Context here: https://github.com/autonomys/subspace/pull/3699#discussion_r2348545020

> // deregistration should not work since the operator is deactivated

Is this actually a property we want?
If we find out that a deactivated operator will never come back, we might want to deregister it from a deactivated state.

Author: Yes.

Staking itself is already complex in terms of Operator state etc..
The change essentially focus on moving operator between deactivation and reactivation(registerd) state

Anything beyond like de-registration etc... would require operator to be registered since the protocol is designed to follow specific state transitions from Registration to deregistration.

So to keep the flow simpler. I would prefer keeping it as is else we will end-up having edge cases where SharePrice might not be calculated

Reviewer: Ok, then let's make a note that a deactivated operator can't be deregistered. This will mean that an operator has to be reactivated to get their stake back (for example).

Reviewer 2: > This will mean that an operator has to be reactivated to get their stake back (for example).

To me, this is problematic and should be addressed in a subsequent phase of implementation.

> offline for any given reason and the block production on Domains is significantly reduced due to the stake deposited for the given operator. By deactivating the operator, the operator status is moved from `Registered` to `Deactiavted` and operator is removed from the Domain operator set and next Operator set along with their stake in the Domain. The rest of the operators will continue producing bundles and expected Domain block production can be restored since the offline operator's stake is not pledged to Domain anymore.

Reviewer 3: ```suggestion
offline for any given reason and the block production on Domains is significantly reduced due to the stake deposited for the given operator. By deactivating the operator, the operator status is moved from `Registered` to `Deactivated` and operator is removed from the Domain operator set and next operator set along with their stake in the Domain. The rest of the operators will continue producing bundles and expected Domain block production can be restored since the offline operator's stake is not pledged to Domain anymore.
```

### reregister_operator

> `reregister_operator(operator_id)`

Reviewer 3: Would this make more sense as `reactivate_operator` or even `activate_operator`? The former might imply a process that would only ever follow a deactivation while the latter might be better if activation is part of (and shares code with) the normal `register_operator` flow.

Author: fair enough. Naming is tough and I went with reregister since the operator status is moved to Register again.
Having said that, I dont mind `activate_operator` here

Reviewer: I think `reactivate_operator` is clearest, because it used to be active. (Activate is easily confused with register / initial registration.)

It is also the clearest opposite to `deactivate_operator`.

> Sudo or Governance can activate a given deactivate operator, given the cooldown period in Epochs has passed for re-registration. Once the operator is activated, the operator is added to the next operator set and will be able to produce bundles from next epoch.

Reviewer: Should there be a way to force reactivate an operator?

I'm thinking of a situation where there is one operator that's working but in the cool-off period, and another that has just gone down, and we only have one or two working operators. In that case we would want to re-register the working operator, even if it was still within its cool off period.

Reviewer 2: I have a couple of more general questions:
1. For governance driven deactivation and re-activation, what is the benefit of the cooldown period?
2. What requirements will need to be met to be re-activated? Should we have a `request_activation` extrinsic?

Author: > For governance driven deactivation and re-activation, what is the benefit of the cooldown period?

Well this cooldown period was intentional even tough it might be redundant at the moment. I have explained why I chose this route in https://github.com/subspace/protocol-specs/pull/70/files#r2351681443

Reviewer: As I said above, I agree with a fixed cooldown period as an experiment to check if it is a useful value (and design).

Still interested in the answer to this question:
> What requirements will need to be met to be re-activated? Should we have a request_activation extrinsic?

Even if the answer is something like:
> > Reactivation is currently manually triggered based on sudo or governance policy, which is out of scope for this specification.

Author: > Sudo or Governance can activate a given deactivate operator, given the cooldown period in Epochs has passed for re-registration.

This is exactly what I have described and I have specifically not included any `force` reactivation as this is purely Sudo/Governance call

> Sudo or Governance can activate a given deactivate operator, given the cooldown period in Epochs has passed for re-registration. Once the operator is activated, the operator is added to the next operator set and will be able to produce bundles from next epoch.

Reviewer 2: ```suggestion
Sudo or Governance can re-activate a given deactivated operator, given the cooldown period in Epochs has passed for re-activation. Once the operator is re-activated, the operator is added to the next operator set and will be able to produce bundles from the next epoch.
```

> Sudo or Governance can reactivate a given deactivate operator, given the `reactivation_delay` period in Epochs has passed for reactivation. Once the operator is activated, the operator is added to the next operator set and will be able to produce bundles from the next epoch.

Reviewer 3: ```suggestion
Sudo or Governance can reactivate a given deactivated operator, given the `reactivation_delay` period in Epochs has passed for reactivation. Once the operator is activated, the operator is added to the next operator set and will be able to produce bundles from the next epoch.
```
